sw/host/opentitanlib/src/crypto/rsa.rs

```diff
@@ -0,0 +1,255 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{anyhow, bail, Result};
+use num_bigint_dig::{traits::ModInverse, BigInt, BigUint, Sign::Minus};
+use rand::rngs::OsRng;
+use rsa::{
+    pkcs8::FromPrivateKey, pkcs8::FromPublicKey, pkcs8::ToPrivateKey, pkcs8::ToPublicKey,
+    PublicKey, PublicKeyParts,
+};
+use std::fs::File;
+use std::io::{Read, Write};
+use std::ops::Deref;
+use std::ops::Shl;
+use std::path::{Path, PathBuf};
+use thiserror::Error;
+
+use crate::crypto::sha256::Sha256Digest;
+use crate::util::bigint::fixed_size_bigint;
+
+const MODULUS_BIT_LEN: usize = 3072;
+const EXPONENT_BIT_LEN: usize = 17;
+const SIGNATURE_BIT_LEN: usize = 3072;
+const RR_BIT_LEN: usize = 3072;
+const N0INV_BIT_LEN: usize = 256;
+
+fixed_size_bigint!(Modulus, MODULUS_BIT_LEN);
+fixed_size_bigint!(Exponent, EXPONENT_BIT_LEN);
+fixed_size_bigint!(Signature, at_most SIGNATURE_BIT_LEN);
+fixed_size_bigint!(RR, at_most RR_BIT_LEN);
+fixed_size_bigint!(N0Inv, at_most N0INV_BIT_LEN);
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Invalid public key")]
+    InvalidPublicKey,
+    #[error("Invalid DER file: {der}")]
+    InvalidDerFile {
+        der: PathBuf,
+        #[source]
+        source: anyhow::Error,
+    },
+    #[error("Read failed: {0}")]
+    ReadFailed(PathBuf),
+    #[error("Write failed: {0}")]
+    WriteFailed(PathBuf),
+    #[error("Generate failed")]
+    GenerateFailed,
+    #[error("Invalid signature")]
+    InvalidSignature,
+    #[error("Sign failed")]
+    SignFailed,
+    #[error("Verification failed")]
+    VerifyFailed,
+    #[error("Failed to compute key component")]
+    KeyComponentComputeFailed,
+}
+
+/// Ensure the components of `key` have the correct bit length.
+fn validate_key(key: impl rsa::PublicKeyParts) -> Result<()> {
+    if key.n().bits() != MODULUS_BIT_LEN || key.e().bits() != EXPONENT_BIT_LEN {
+        bail!(Error::InvalidPublicKey)
+    } else {
+        Ok(())
+    }
+}
```

Comment:
nit: `Modulus` and `Exponent` already expect exactly 3072 and 17 bits, respectively. Can we use those types in the constructors instead? Also, I think we should also check that exponent is 65537.

Reply:
SG, I misremembered the conclusion that we reached. Thanks!